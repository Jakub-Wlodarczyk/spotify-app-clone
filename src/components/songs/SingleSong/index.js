import React from 'react';
import { addSongToPlaylist } from '../../../store/actions/playlists';
import { savePlaylistsState } from '../../../resources/api';

class SingleSong extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			showPlaylists: false,
			currentPlaylist: '',
			showConfirm: false
		};

		this.handlePlaylistsView = this.handlePlaylistsView.bind(this);
		this.handlePlaylistChange = this.handlePlaylistChange.bind(this);
		this.handleSongAdding = this.handleSongAdding.bind(this);
	}

	componentDidUpdate() {
		savePlaylistsState(this.props.playlists);
	}

	handlePlaylistsView() {
		this.setState((prevState) => ({
			showPlaylists: !prevState.showPlaylists,
			showConfirm: true
		}));
	}

	handlePlaylistChange(e) {
		const playlist = e.target.value;
		if (playlist) {
			this.setState({
				currentPlaylist: playlist
			});
		}
	}

	handleSongAdding() {
		if (this.state.currentPlaylist) {
			this.props.dispatch(addSongToPlaylist({
				songName: this.props.song.title,
				playlistId: this.state.currentPlaylist
			}));
		}
	}

	render() {
		const { title, author, duration } = this.props.song;
		const { playlists } = this.props;
		const { showPlaylists, currentPlaylist, showConfirm } = this.state;

		return (
			<div className="single-song">
				<p><b>Song title:</b> {title}</p>
				<p><b>Song author:</b> {author}</p>
				<p><b>Song duration:</b> {duration}</p>
				<button type="button" onClick={this.handlePlaylistsView}>Add To Playlist</button>

				{
					showPlaylists &&
						<select value={currentPlaylist} onChange={this.handlePlaylistChange}>
							<option value="">select playlist</option>
							{playlists.map(playlist => <option key={playlist.id} value={playlist.id}>{playlist.name}</option>)}
						</select>
				}
				{
					(playlists.length > 0 && showConfirm && showPlaylists) && <button type="button" onClick={this.handleSongAdding}>Confirm adding</button>
				}
			</div>
		);
	}
}

export default SingleSong;
