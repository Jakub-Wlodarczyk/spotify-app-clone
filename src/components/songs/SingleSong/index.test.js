import React from 'react';
import { shallow } from 'enzyme';
import SingleSong from './index';

describe('SingleSong component', () => {
	const props = {
		playlists: [{ name: 'playlist', songs: ['song1', 'song2']}],
		song: { title: 'title 1', author: 'author 1', duration: '4:30' }
	};
	it('should match the snapshot', () => {
		const wrapper = shallow(<SingleSong {...props} />);
		expect(wrapper).toMatchSnapshot();
	});
});
