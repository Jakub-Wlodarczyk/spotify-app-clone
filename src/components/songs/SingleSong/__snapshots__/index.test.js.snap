// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleSong component should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SingleSong
    playlists={
      Array [
        Object {
          "name": "playlist",
          "songs": Array [
            "song1",
            "song2",
          ],
        },
      ]
    }
    song={
      Object {
        "author": "author 1",
        "duration": "4:30",
        "title": "title 1",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p>
          <b>
            Song title:
          </b>
           
          title 1
        </p>,
        <p>
          <b>
            Song author:
          </b>
           
          author 1
        </p>,
        <p>
          <b>
            Song duration:
          </b>
           
          4:30
        </p>,
        <button
          onClick={[Function]}
          type="button"
        >
          Add To Playlist
        </button>,
        false,
        false,
      ],
      "className": "single-song",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <b>
              Song title:
            </b>,
            " ",
            "title 1",
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Song title:",
            },
            "ref": null,
            "rendered": "Song title:",
            "type": "b",
          },
          " ",
          "title 1",
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <b>
              Song author:
            </b>,
            " ",
            "author 1",
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Song author:",
            },
            "ref": null,
            "rendered": "Song author:",
            "type": "b",
          },
          " ",
          "author 1",
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <b>
              Song duration:
            </b>,
            " ",
            "4:30",
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Song duration:",
            },
            "ref": null,
            "rendered": "Song duration:",
            "type": "b",
          },
          " ",
          "4:30",
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Add To Playlist",
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": "Add To Playlist",
        "type": "button",
      },
      false,
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p>
            <b>
              Song title:
            </b>
             
            title 1
          </p>,
          <p>
            <b>
              Song author:
            </b>
             
            author 1
          </p>,
          <p>
            <b>
              Song duration:
            </b>
             
            4:30
          </p>,
          <button
            onClick={[Function]}
            type="button"
          >
            Add To Playlist
          </button>,
          false,
          false,
        ],
        "className": "single-song",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <b>
                Song title:
              </b>,
              " ",
              "title 1",
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Song title:",
              },
              "ref": null,
              "rendered": "Song title:",
              "type": "b",
            },
            " ",
            "title 1",
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <b>
                Song author:
              </b>,
              " ",
              "author 1",
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Song author:",
              },
              "ref": null,
              "rendered": "Song author:",
              "type": "b",
            },
            " ",
            "author 1",
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <b>
                Song duration:
              </b>,
              " ",
              "4:30",
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Song duration:",
              },
              "ref": null,
              "rendered": "Song duration:",
              "type": "b",
            },
            " ",
            "4:30",
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add To Playlist",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": "Add To Playlist",
          "type": "button",
        },
        false,
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
